File: 0443-string-compression/0443-string-compression.js
var compress = function (chars) {
    let write = 0;
    let count = 1
    let val = chars[0]

    function makeArr(count, val) {
        chars[write] = val;
        write++;
        if (count !== 1) {
            let s = count.toString()
            for (let j = 0; j < s.length; j++) {
                chars[write] = s[j];
                write++
            }
        }
    }

    for (let i = 1; i < chars.length; i++) {
        if (val === chars[i]) {
            count++
        } else {
            makeArr(count, val)
            val = chars[i]
            count = 1
        }
    }

    makeArr(count, val)

    chars.length = write

    return chars.length
};

File: 0443-string-compression/0443-string-compression.cs
public class Solution
{
    public int Compress(char[] chars)
    {
        int write = 0;
        char val = chars[0];
        int count = 1;

        void makeArr(int count, char val)
        {
            chars[write++] = val;

            if (count != 1)
            {
                string s = count.ToString();
                foreach (char c in s)
                {
                    chars[write++] = c;
                }
            }
        }

        for (int i = 1; i < chars.Length; i++)
        {
            if (val == chars[i])
            {
                count++;
            }
            else
            {
                makeArr(count, val);
                val = chars[i];
                count = 1;
            }
        }

        makeArr(count, val);

        return write;
    }
}
